### DISCUSSION TITLE
How to change generated project using composer plugin and `POST_CREATE_PROJECT_CMD` event

### DISCUSSION THREAD

**Discussion by JurrienLetsTalk** [composer/composer#12025] [Q&A / Support] [answered]
I am trying to change a generated composer project by supplying a composer plugin in the template. 

When running the `composer create-project <template>` command, this plugin gets executed with the following steps:

1. Plugin asks for new package name
2. Plugin alters `composer.json` of the new project with the new package name
3. Plugin removes itself from the new project.

I can not unify the last 2 steps however.

I have currently two (related) problems;

1. I do not know how to change the new project name in the `composer.json` file without directly altering the file
2. I am using the following code to remove the plugin package, but this does not seem to work.

```php
$localRepository = $this->composer->getRepositoryManager()->getLocalRepository();
$ownPackage      = $localRepository->findPackage($this->packageName, '*');
$this->composer->getInstallationManager()->uninstall($localRepository, new UninstallOperation($ownPackage));
```

After the plugin is done, the changes directly written to the file by way of `file_put_contents` are seen, but the removal of the plugin package is not.

Obviously I could simply execute the composer command in the shell, but that seems counterintuitive. 

What am I doing wrong here? Or how can I change the created project's package name?

**Comment by stof** [comment, marked as answer]
See https://github.com/symfony/flex/blob/4dc11919791f81d087a12db2ab4c7e044431ef6b/src/Flex.php#L242-L271 for an example of how symfony/flex modifies the root composer.json after creating a project.

Regarding uninstalling the plugin, you would have to modify the requirements in the composer.json and run the solver again to have a proper removal. Using the installation manager will uninstall the plugin from the vendor folder, but won't remove it from the lock file.
